src/aiohabit/outputs/pytest_multihost.py

```diff
@@ -0,0 +1,128 @@
+        # topology config in metadata contains more values (os, group) than the ones
+        # needed by pytest multihost. They need to be removed in order to work.
+        host_allowed_attrs = [
+            "name",
+            "role",
+            "hostname",
+            "shortname",
+            "external_hostname",
+            "ip",
+            "domain",
+            "username",
+            "password",
+            "host_type",
+        ]
+        domain_allowed_attrs = ["name", "type", "hosts"]
+
+        # Use values from provisioning-config mhcfg section - as default values
+        for option in self._config["mhcfg"]:
+            # options defined in metadata takes precedence over options in
+            # provisioning-config as it is "closer" to user.
+            if option in mhcfg:
+                continue
+            mhcfg[option] = self._config["mhcfg"][option]
+
+        for domain in mhcfg["domains"]:
+            for host in domain["hosts"]:
+                provisioned_host = self._db.hosts[host["name"]]
+                username = get_val([provisioned_host.username, host.get("username")])
+                password = get_val([provisioned_host.password, host.get("password")])
+
+                if is_windows_host(host):
+                    host["password"] = get_val(
+                        [password, mhcfg.get("ad_admin_password")]
+                    )
+                    host["username"] = get_val([username, "Administrator"])
+                else:
+                    if username:
+                        host["username"] = username
+
+                ip = provisioned_host.ips[0]
+                dns_record = resolve_hostname(ip)
+                host["ip"] = ip
+                host["external_hostname"] = dns_record
+
+                if host["group"] == "ad_root":
+                    mhcfg["ad_top_domain"] = domain["name"]
+                    mhcfg["ad_hostname"] = host["name"]
+                    mhcfg["ad_ip"] = host["ip"]
+                elif host["group"] == "ad_subdomain":
+                    mhcfg["ad_sub_domain"] = domain["name"]
+                    mhcfg["ad_sub_hostname"] = host["name"]
+                    mhcfg["ad_sub_ip"] = host["ip"]
+
+                rm_keys = [key for key in host.keys() if key not in host_allowed_attrs]
+                for key in rm_keys:
+                    del host[key]
+
+            domain_rm_keys = [
+                key for key in domain.keys() if key not in domain_allowed_attrs
+            ]
+            for key in domain_rm_keys:
+                del domain[key]
+
+        # ssh_key_filename must be absolute as it can be used from a different
+        # working directory, e.g. running tests from git repo
+        ssh_key_filename = mhcfg.get("ssh_key_filename")
+        if ssh_key_filename and not ssh_key_filename.startswith("~"):
+            mhcfg["ssh_key_filename"] = os.path.abspath(ssh_key_filename)
+
+        return mhcfg
+
+    def create_output(self):
```

Comment:
I guess this will be the only common method used by all "output classes", right?

Reply:
That was the intend, but we don't need to stick with it if it will not make sense.